core/src/cluster_info_repair_listener.rs

```diff
@@ -0,0 +1,683 @@
+                        // Loop over the blob indexes and query the database for these blob that
+                        // this node is reponsible for repairing. This should be faster than using
+                        // a database iterator over the slots because by the time this node is
+                        // sending the blobs in this slot for repair, we expect these slots
+                        // to be full.
+                        let blob_index = (start_index + i) % num_blobs_in_slot;
+                        if let Some(blob_data) = blocktree
+                            .get_data_blob_bytes(slot, blob_index as u64)
+                            .expect("Failed to read data blob from blocktree")
+                        {
+                            socket.send_to(&blob_data[..], repairee_tvu)?;
+                            total_data_blobs_sent += 1;
+                            should_sleep = true;
+                        }
+
+                        if let Some(coding_bytes) = blocktree
+                            .get_coding_blob_bytes(slot, blob_index as u64)
+                            .expect("Failed to read coding blob from blocktree")
+                        {
+                            socket.send_to(&coding_bytes[..], repairee_tvu)?;
+                            total_coding_blobs_sent += 1;
+                            should_sleep = true;
+                        }
+
+                        if should_sleep {
+                            sleep(Duration::from_millis(BLOB_SEND_SLEEP_MILLIS as u64));
+                        }
+                    }
+                }
+            }
+        }
+
+        Self::report_repair_metrics(total_data_blobs_sent, total_coding_blobs_sent);
+        Ok(())
+    }
+
+    fn report_repair_metrics(total_data_blobs_sent: u64, total_coding_blobs_sent: u64) {
+        if total_data_blobs_sent > 0 || total_coding_blobs_sent > 0 {
+            datapoint!(
+                "repairman_activity",
+                ("data_sent", total_data_blobs_sent, i64),
+                ("coding_sent", total_coding_blobs_sent, i64)
+            );
+        }
+    }
+
+    fn shuffle_repairmen(
+        eligible_repairmen: &mut Vec<&Pubkey>,
+        repairee_id: &Pubkey,
+        repairee_root: u64,
+    ) {
+        // Make a seed from pubkey + repairee root
+        let mut seed = [0u8; mem::size_of::<Pubkey>()];
+        let repairee_id_bytes = repairee_id.as_ref();
+        seed[..repairee_id_bytes.len()].copy_from_slice(repairee_id_bytes);
+        LittleEndian::write_u64(&mut seed[0..], repairee_root);
+
+        // Deterministically shuffle the eligible repairmen based on the seed
+        let mut rng = ChaChaRng::from_seed(seed);
+        eligible_repairmen.shuffle(&mut rng);
+    }
+
+    // Returns (begin_index, num_blobs), signaling that the repairman is reponsible
+    // for blob indexes in the range [begin_index, begin_index + num_blobs] for this
```

Comment:
I think it'd be a bunch better to have repairmen responsible for every Nth rather than a particular range, that should spread out the traditional repairs, make erasure recovery more likely